hal/hal_remote/source/hal_socket_transmitter.cpp

```diff
@@ -0,0 +1,220 @@
+    return false;
+  }
+  last_error = hal_socket_transmitter::success;
+  return true;
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::connect() {
+  const int res = ::connect(sock, (struct sockaddr *)&server_address,
+                            sizeof(server_address));
+  if (res != -1) {
+    fd_to_use = sock;
+    is_connected = true;
+    return hal_socket_transmitter::success;
+  } else {
+    return hal_socket_transmitter::connect_failed;
+  }
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::setup_connection(
+    bool server) {
+  // don't support port 0
+  if (port_requested == 0) {
+    if (debug_enabled()) {
+      (void)fprintf(stderr, "port 0 requested. This is disallowed\n");
+    }
+    return hal_socket_transmitter::port_0_requested;
+  }
+  sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+
+  struct addrinfo hints;
+  struct addrinfo *result;
+  int sfd, s;
+  std::memset(&hints, 0, sizeof(struct addrinfo));
+  hints.ai_family = AF_UNSPEC; /* Allow IPv4 or IPv6 */
+  hints.ai_socktype = SOCK_STREAM;
+  hints.ai_flags = AI_NUMERICSERV;
+  hints.ai_protocol = IPPROTO_TCP;
+
+  const std::string port_str = std::to_string(port_requested);
+  s = getaddrinfo(node.c_str(), port_str.c_str(), &hints, &result);
+  // Use getaddrinfo on the node and port. This may return one or
+  // more entries we can bind to in priority order. We take the first
+  // one that matches.
+  if (s != 0) {
+    if (debug_enabled()) {
+      (void)fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s));
+    }
+    return hal_socket_transmitter::getaddrinfo_failed;
+  } else {
+    bool bind_failed = false;
+    bool socket_failed = false;
+    for (auto *rp = result; rp != NULL; rp = rp->ai_next) {
+      bind_failed = false;
+      socket_failed = false;
+      sock = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+      if (sock == -1) {
+        socket_failed = true;
+        continue;
+      }
+      server_address = *rp->ai_addr;
+      if (server) {
+        if (const int res = bind(sock, rp->ai_addr, rp->ai_addrlen) != 0) {
+          close(sock);
+          sock = -1;
+          bind_failed = true;
+          continue;
+        }
+
+        struct sockaddr_in local_addr;
+        socklen_t len = sizeof(local_addr);
+        if (const int res =
+                getsockname(sock, (struct sockaddr *)&local_addr, &len) != 0) {
+          return hal_socket_transmitter::getsockname_failed;
+        }
+        current_port = ntohs(local_addr.sin_port);
+      }
+    }
+    if (sock == -1) {
+      if (bind_failed) {
+        return hal_socket_transmitter::bind_failed;
+      } else if (socket_failed) {
+        return hal_socket_transmitter::socket_failed;
+      }
+    }
+    if (!server) {
+      current_port = port_requested;
+    }
+  }
+  setup_connection_done = true;
+  return hal_socket_transmitter::success;
+}
+
+int hal_socket_transmitter::listen() {
+  if (const int res = ::listen(sock, 1) != 0) {
```

Comment:
This means `res` is initialised to `::listen(sock, 1) != 0`, i.e. it will always be zero or one, rather than returning the result of `::listen(sock, 1)`. Also, the `if (res != 0) return res; else return 0;` logic looks overly complicated, that amounts to just `return res;`.

But given that the API here is "1 is failure, 0 is success", whereas below for `accept()`, it's "-1 is failure, anything else, possibly zero, is the file descriptor that must not be used", I wonder if it's worth updating both to just return `bool`?

Reply:
This pattern of `const int res = ... != 0` appears in a few other places where that does not look intentional.